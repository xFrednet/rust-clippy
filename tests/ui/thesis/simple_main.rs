//@rustc-env: CLIPPY_PRINT_MIR=1

#![allow(unused)]

fn magic_1<T>(b: &T) {}
fn magic_2<T>(b: &T, c: &T) {}

fn print_mir() {
    let a = if true {
        let mut x = Vec::new();
        x.push(1);
        x
    } else {
        let mut y = Vec::new();
        y.push(89);
        y
    };
}

struct A {
    field: String,
}

impl A {
    fn borrow_field_direct(&self) -> &String {
        &self.field
    }

    fn borrow_field_deref(&self) -> &str {
        &self.field
    }
}

// fn simple_ownership(cond: bool) {
//     let a = String::new();
//     let b = String::new();
//
//     let x;
//     if cond {
//         x = &a;
//     } else {
//         x = &b;
//     }
//
//     magic_1(x);
// }

// fn if_fun(a: String, b: String, cond: bool) {
//     if cond {
//         magic_1(&a);
//     } else {
//         magic_1(&b);
//     }
//
//     magic_1(if cond {&a} else {&b});
// }

fn main() {}
