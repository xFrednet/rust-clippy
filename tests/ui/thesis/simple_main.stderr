thread 'rustc' panicked at clippy_lints/src/borrow_pats.rs:332:39:
not yet implemented: Automata {
    local: _2,
    local_kind: AnonVar,
    body: Body {
        basic_blocks: BasicBlocks {
            basic_blocks: [
                BasicBlockData {
                    statements: [
                        StorageLive(_2),
                        StorageLive(_3),
                        _3 = &_1,
                    ],
                    terminator: Some(
                        Terminator {
                            source_info: SourceInfo {
                                span: tests/ui/thesis/simple_main.rs:35:8: 35:24 (#0),
                                scope: scope[0],
                            },
                            kind: _2 = std::vec::Vec::<i32>::is_empty(move _3) -> [return: bb1, unwind: bb7],
                        },
                    ),
                    is_cleanup: false,
                },
                BasicBlockData {
                    statements: [],
                    terminator: Some(
                        Terminator {
                            source_info: SourceInfo {
                                span: tests/ui/thesis/simple_main.rs:35:8: 35:24 (#0),
                                scope: scope[0],
                            },
                            kind: switchInt(move _2) -> [0: bb4, otherwise: bb2],
                        },
                    ),
                    is_cleanup: false,
                },
                BasicBlockData {
                    statements: [
                        StorageDead(_3),
                        StorageLive(_4),
                        StorageLive(_5),
                        _5 = &mut _1,
                    ],
                    terminator: Some(
                        Terminator {
                            source_info: SourceInfo {
                                span: tests/ui/thesis/simple_main.rs:36:9: 36:22 (#0),
                                scope: scope[0],
                            },
                            kind: _4 = std::vec::Vec::<i32>::clear(move _5) -> [return: bb3, unwind: bb7],
                        },
                    ),
                    is_cleanup: false,
                },
                BasicBlockData {
                    statements: [
                        StorageDead(_5),
                        StorageDead(_4),
                        _0 = const (),
                    ],
                    terminator: Some(
                        Terminator {
                            source_info: SourceInfo {
                                span: tests/ui/thesis/simple_main.rs:35:5: 37:6 (#0),
                                scope: scope[0],
                            },
                            kind: goto -> bb5,
                        },
                    ),
                    is_cleanup: false,
                },
                BasicBlockData {
                    statements: [
                        StorageDead(_3),
                        _0 = const (),
                    ],
                    terminator: Some(
                        Terminator {
                            source_info: SourceInfo {
                                span: tests/ui/thesis/simple_main.rs:35:5: 37:6 (#0),
                                scope: scope[0],
                            },
                            kind: goto -> bb5,
                        },
                    ),
                    is_cleanup: false,
                },
                BasicBlockData {
                    statements: [
                        StorageDead(_2),
                    ],
                    terminator: Some(
                        Terminator {
                            source_info: SourceInfo {
                                span: tests/ui/thesis/simple_main.rs:38:1: 38:2 (#0),
                                scope: scope[0],
                            },
                            kind: drop(_1) -> [return: bb6, unwind: bb8],
                        },
                    ),
                    is_cleanup: false,
                },
                BasicBlockData {
                    statements: [],
                    terminator: Some(
                        Terminator {
                            source_info: SourceInfo {
                                span: tests/ui/thesis/simple_main.rs:38:2: 38:2 (#0),
                                scope: scope[0],
                            },
                            kind: return,
                        },
                    ),
                    is_cleanup: false,
                },
                BasicBlockData {
                    statements: [],
                    terminator: Some(
                        Terminator {
                            source_info: SourceInfo {
                                span: tests/ui/thesis/simple_main.rs:38:1: 38:2 (#0),
                                scope: scope[0],
                            },
                            kind: drop(_1) -> [return: bb8, unwind terminate(cleanup)],
                        },
                    ),
                    is_cleanup: true,
                },
                BasicBlockData {
                    statements: [],
                    terminator: Some(
                        Terminator {
                            source_info: SourceInfo {
                                span: tests/ui/thesis/simple_main.rs:34:1: 38:2 (#0),
                                scope: scope[0],
                            },
                            kind: resume,
                        },
                    ),
                    is_cleanup: true,
                },
            ],
            cache: Cache {
                predecessors: OnceLock(
                    <uninit>,
                ),
                switch_sources: OnceLock(
                    <uninit>,
                ),
                is_cyclic: OnceLock(
                    <uninit>,
                ),
                reverse_postorder: OnceLock(
                    <uninit>,
                ),
                dominators: OnceLock(
                    <uninit>,
                ),
            },
        },
        phase: Analysis(
            Initial,
        ),
        pass_count: 1,
        source: MirSource {
            instance: Item(
                DefId(0:13 ~ simple_main[0912]::simple_ownership),
            ),
            promoted: None,
        },
        source_scopes: [
            SourceScopeData {
                span: tests/ui/thesis/simple_main.rs:34:1: 38:2 (#0),
                parent_scope: None,
                inlined: None,
                inlined_parent_scope: None,
                local_data: Set(
                    SourceScopeLocalData {
                        lint_root: HirId(DefId(0:13 ~ simple_main[0912]::simple_ownership).0),
                        safety: Safe,
                    },
                ),
            },
        ],
        coroutine: None,
        local_decls: [
            LocalDecl {
                mutability: Mut,
                local_info: Set(
                    Boring,
                ),
                ty: (),
                user_ty: None,
                source_info: SourceInfo {
                    span: tests/ui/thesis/simple_main.rs:34:41: 34:41 (#0),
                    scope: scope[0],
                },
            },
            LocalDecl {
                mutability: Mut,
                local_info: Set(
                    User(
                        Var(
                            VarBindingForm {
                                binding_mode: BindByValue(
                                    Mut,
                                ),
                                opt_ty_info: Some(
                                    tests/ui/thesis/simple_main.rs:34:32: 34:40 (#0),
                                ),
                                opt_match_place: Some(
                                    (
                                        None,
                                        tests/ui/thesis/simple_main.rs:34:21: 34:30 (#0),
                                    ),
                                ),
                                pat_span: tests/ui/thesis/simple_main.rs:34:21: 34:30 (#0),
                            },
                        ),
                    ),
                ),
                ty: std::vec::Vec<i32, std::alloc::Global>,
                user_ty: None,
                source_info: SourceInfo {
                    span: tests/ui/thesis/simple_main.rs:34:21: 34:30 (#0),
                    scope: scope[0],
                },
            },
            LocalDecl {
                mutability: Mut,
                local_info: Set(
                    BlockTailTemp(
                        BlockTailInfo {
                            tail_result_is_ignored: true,
                            span: tests/ui/thesis/simple_main.rs:35:5: 37:6 (#0),
                        },
                    ),
                ),
                ty: bool,
                user_ty: None,
                source_info: SourceInfo {
                    span: tests/ui/thesis/simple_main.rs:35:8: 35:24 (#0),
                    scope: scope[0],
                },
            },
            LocalDecl {
                mutability: Mut,
                local_info: Set(
                    BlockTailTemp(
                        BlockTailInfo {
                            tail_result_is_ignored: true,
                            span: tests/ui/thesis/simple_main.rs:35:5: 37:6 (#0),
                        },
                    ),
                ),
                ty: &ReErased std::vec::Vec<i32, std::alloc::Global>,
                user_ty: None,
                source_info: SourceInfo {
                    span: tests/ui/thesis/simple_main.rs:35:8: 35:13 (#0),
                    scope: scope[0],
                },
            },
            LocalDecl {
                mutability: Not,
                local_info: Set(
                    Boring,
                ),
                ty: (),
                user_ty: None,
                source_info: SourceInfo {
                    span: tests/ui/thesis/simple_main.rs:36:9: 36:22 (#0),
                    scope: scope[0],
                },
            },
            LocalDecl {
                mutability: Mut,
                local_info: Set(
                    Boring,
                ),
                ty: &ReErased mut std::vec::Vec<i32, std::alloc::Global>,
                user_ty: None,
                source_info: SourceInfo {
                    span: tests/ui/thesis/simple_main.rs:36:9: 36:14 (#0),
                    scope: scope[0],
                },
            },
        ],
        user_type_annotations: [],
        arg_count: 1,
        spread_arg: None,
        var_debug_info: [
            owned => _1,
        ],
        span: tests/ui/thesis/simple_main.rs:34:1: 38:2 (#0),
        required_consts: [],
        is_polymorphic: false,
        injection_phase: None,
        tainted_by_errors: None,
        function_coverage_info: None,
    },
    state: Owned(
        Empty,
    ),
    events: [],
    best_pet: Unknown,
}

Event {
    bb: bb1,
    local: _2,
    kind: Move(
        Switch,
    ),
}
stack backtrace:
   0:     0x7bca7cb8caf6 - std::backtrace_rs::backtrace::libunwind::trace::hf298c77755f5b12e
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:     0x7bca7cb8caf6 - std::backtrace_rs::backtrace::trace_unsynchronized::h4da6a489ce2db74e
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7bca7cb8caf6 - std::sys_common::backtrace::_print_fmt::h051beca4a6f5cb98
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x7bca7cb8caf6 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hcb86488900881a45
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7bca7cbdda7c - core::fmt::rt::Argument::fmt::haf43875fa4a971bb
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/fmt/rt.rs:142:9
   5:     0x7bca7cbdda7c - core::fmt::write::h44c7f32ae3be5535
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/fmt/mod.rs:1120:17
   6:     0x7bca7cb814bf - std::io::Write::write_fmt::hc3fec0223f64c010
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/io/mod.rs:1846:15
   7:     0x7bca7cb8c8a4 - std::sys_common::backtrace::_print::h7c03376d42fd745a
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7bca7cb8c8a4 - std::sys_common::backtrace::print::h9ff10455f68b06c1
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7bca7cb8f5eb - std::panicking::default_hook::{{closure}}::he4f9865f1b72b2d5
  10:     0x7bca7cb8f339 - std::panicking::default_hook::h3d59f13a05d49fe6
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:292:9
  11:     0x7bca7fa731ec - std[19d33715048712a6]::panicking::update_hook::<alloc[6fc58ab69ba746fe]::boxed::Box<rustc_driver_impl[4c01069949ad96dd]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7bca7cb8fd50 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hf41c577e6ac4b0df
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/alloc/src/boxed.rs:2030:9
  13:     0x7bca7cb8fd50 - std::panicking::rust_panic_with_hook::h70f59ab226ae2c5c
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:786:13
  14:     0x7bca7cb8fa92 - std::panicking::begin_panic_handler::{{closure}}::h47fc7126fa960711
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:660:13
  15:     0x7bca7cb8cfd6 - std::sys_common::backtrace::__rust_end_short_backtrace::h5b94f863278adeaf
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x7bca7cb8f7c4 - rust_begin_unwind
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/panicking.rs:648:5
  17:     0x7bca7cbd9f95 - core::panicking::panic_fmt::h96a33d9be4ce8c0f
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/core/src/panicking.rs:72:14
  18:     0x6260ae60ece3 - clippy_lints::borrow_pats::Automata::update_owned_state::h883ac80defed3898
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats.rs:332:39
  19:     0x6260ae60e8b8 - clippy_lints::borrow_pats::Automata::accept_event::h49bf97a60bb4ac95
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats.rs:316:32
  20:     0x6260ae60e025 - clippy_lints::borrow_pats::BorrowAnalysis::accept_event::he23edcd375526e86
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats.rs:195:20
  21:     0x6260ae60e53e - clippy_lints::borrow_pats::BorrowAnalysis::do_term::h6aa1d7be20fbc327
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats.rs:235:25
  22:     0x6260ae60db71 - clippy_lints::borrow_pats::BorrowAnalysis::do_body::hc5f769b0aafd3241
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats.rs:134:24
  23:     0x6260ae60f74e - <clippy_lints::borrow_pats::BorrowPats as rustc_lint::passes::LateLintPass>::check_body::hf3889422ad231c8d
                               at /home/xfrednet/workspace/rust/rust-clippy/clippy_lints/src/borrow_pats.rs:504:13
  24:     0x7bca7fdfa83b - <rustc_lint[a04349e93729da6f]::late::LateContextAndPass<rustc_lint[a04349e93729da6f]::late::RuntimeCombinedLateLintPass> as rustc_hir[9ace93772c87fb15]::intravisit::Visitor>::visit_nested_body
  25:     0x7bca7fdfb6e4 - <rustc_lint[a04349e93729da6f]::late::LateContextAndPass<rustc_lint[a04349e93729da6f]::late::RuntimeCombinedLateLintPass> as rustc_hir[9ace93772c87fb15]::intravisit::Visitor>::visit_fn
  26:     0x7bca7fdeaae4 - <rustc_lint[a04349e93729da6f]::late::LateContextAndPass<rustc_lint[a04349e93729da6f]::late::RuntimeCombinedLateLintPass> as rustc_hir[9ace93772c87fb15]::intravisit::Visitor>::visit_nested_item
  27:     0x7bca81a2d473 - rustc_lint[a04349e93729da6f]::late::check_crate::{closure#0}
  28:     0x7bca81a2d6d1 - rustc_lint[a04349e93729da6f]::late::check_crate
  29:     0x7bca818a39c9 - rustc_interface[3868b2f10de3354]::passes::analysis
  30:     0x7bca818a2669 - rustc_query_impl[80d5b36ddf5440ab]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[80d5b36ddf5440ab]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[3375709bf24a03bb]::query::erase::Erased<[u8; 1usize]>>
  31:     0x7bca81b33ce5 - rustc_query_system[3c4b6c562f3523cb]::query::plumbing::try_execute_query::<rustc_query_impl[80d5b36ddf5440ab]::DynamicConfig<rustc_query_system[3c4b6c562f3523cb]::query::caches::SingleCache<rustc_middle[3375709bf24a03bb]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[80d5b36ddf5440ab]::plumbing::QueryCtxt, false>
  32:     0x7bca81b33a49 - rustc_query_impl[80d5b36ddf5440ab]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7bca81b42f59 - rustc_interface[3868b2f10de3354]::interface::run_compiler::<core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>, rustc_driver_impl[4c01069949ad96dd]::run_compiler::{closure#0}>::{closure#0}
  34:     0x7bca81dbdc45 - std[19d33715048712a6]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[3868b2f10de3354]::util::run_in_thread_with_globals<rustc_interface[3868b2f10de3354]::util::run_in_thread_pool_with_globals<rustc_interface[3868b2f10de3354]::interface::run_compiler<core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>, rustc_driver_impl[4c01069949ad96dd]::run_compiler::{closure#0}>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>
  35:     0x7bca81dbda72 - <<std[19d33715048712a6]::thread::Builder>::spawn_unchecked_<rustc_interface[3868b2f10de3354]::util::run_in_thread_with_globals<rustc_interface[3868b2f10de3354]::util::run_in_thread_pool_with_globals<rustc_interface[3868b2f10de3354]::interface::run_compiler<core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>, rustc_driver_impl[4c01069949ad96dd]::run_compiler::{closure#0}>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[a9f526eb09c6e7cd]::result::Result<(), rustc_span[a2e62ba9c1d34a9a]::ErrorGuaranteed>>::{closure#1} as core[a9f526eb09c6e7cd]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  36:     0x7bca7cb99725 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h79a9e4aa91d71737
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/alloc/src/boxed.rs:2016:9
  37:     0x7bca7cb99725 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h6eabe626cfdc74a5
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/alloc/src/boxed.rs:2016:9
  38:     0x7bca7cb99725 - std::sys::pal::unix::thread::Thread::new::thread_start::h8428b71c070d7cf2
                               at /rustc/3406ada96f8e16e49e947a91db3eba0db45245fa/library/std/src/sys/pal/unix/thread.rs:108:17
  39:     0x7bca75094ac3 - start_thread
                               at ./nptl/pthread_create.c:442:8
  40:     0x7bca75126850 - __GI___clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
  41:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new?template=ice.yml

note: please attach the file at `/home/xfrednet/workspace/rust/rust-clippy/rustc-ice-2024-03-26T16_39_31-1716539.txt` to your bug report

note: compiler flags: -Z ui-testing -Z deduplicate-diagnostics=no

query stack during panic:
#0 [analysis] running analysis passes on this crate
end of query stack
note: Clippy version: clippy 0.1.78 (56f1cdc61 2024-03-25)

